src/main/java/com/rcv/TabulatorSession.java

```diff
@@ -0,0 +1,260 @@
+          configPath);
+    } else if (rawConfig.cvrFileSources == null) {
+      Logger.log(
+          Level.SEVERE,
+          "No 'cvrFileSources' field specified! Failed to load config file:\n%s",
+          configPath);
+    } else if (rawConfig.candidates == null) {
+      Logger.log(
+          Level.SEVERE,
+          "No 'candidates' field specified! Failed to load config file:\n%s",
+          configPath);
+    } else if (rawConfig.rules == null) {
+      Logger.log(
+          Level.SEVERE, "No 'rules' field specified! Failed to load config file:\n%s", configPath);
+    } else {
+      // proceed to create the ContestConfig wrapper
+      config = new ContestConfig(rawConfig);
+      Logger.log(Level.INFO, "Successfully loaded config file: %s", configPath);
+    }
+    return config;
+  }
+
+  // function: executeTabulation
+  // purpose: execute tabulation for given ContestConfig
+  // param: config object containing CVR file paths to parse
+  // returns: String indicating whether or not execution was successful
+  private void executeTabulation(ContestConfig config) {
+    Logger.log(Level.INFO, "Starting tabulation process...");
+
+    boolean isTabulationCompleted = false;
+    boolean isConfigValid = config.validate();
+
+    if (isConfigValid) {
+      boolean isTabulationLogSetUp = false;
+      // current date-time formatted as a string used for creating unique output files names
+      final String timestampString = new SimpleDateFormat("yyyy-MM-dd_HH-mm-ss").format(new Date());
+      // %g format is for log file naming
+      String tabulationLogPath =
+          Paths.get(config.getOutputDirectory(), String.format("%s_audit_%%g.log",
+              timestampString))
+              .toAbsolutePath()
+              .toString();
+
+      // cache summaryOutputPath for testing
+      summaryOutputPath =
+          Paths.get(config.getOutputDirectory(), String.format("%s_summary.json",
+              timestampString))
+              .toAbsolutePath()
+              .toString();
+      try {
+        FileUtils.createOutputDirectory(config.getOutputDirectory());
+        Logger.addTabulationFileLogging(tabulationLogPath);
+        isTabulationLogSetUp = true;
+      } catch (UnableToCreateDirectoryException exception) {
+        Logger.log(
+            Level.SEVERE,
+            "Failed to create output directory: %s\n%s",
+            config.getOutputDirectory(),
+            exception.toString());
+      } catch (IOException exception) {
+        Logger.log(
+            Level.SEVERE, "Failed to configure tabulation logger!\n%s", exception.toString());
+      }
+
+      if (isTabulationLogSetUp) {
+        Logger.log(Level.INFO, "Logging tabulation to: %s", tabulationLogPath);
+        // Read cast vote records from CVR files
+        // parse the cast vote records
+        Pair<List<CastVoteRecord>, Set<String>> cvrInfo = parseCastVoteRecords(config);
```

Comment:
Not a big deal, but it's unfortunate that we then have to call `getValue` and `getKey` on this when it's not actually a key-value pair...

Reply:
I changed it to pass the IDs as a parameter so it's a bit simpler.